### PR TITLE
feat: free tx staking

### PR DESCRIPTION
**Pull request by mtwaro**

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -2295,13 +2304,15 @@ bool CChainState::ConnectBlock(const CBlock& block, BlockValidationState& state,
 
             // TODO: double parsing of the transaction. Should be done more efficiently.
             if (fStakingActive) {
+                CStakingTransactionParser stakingTxParser(MakeTransactionRef(tx));
                 if (txfee == 0) {
-                    if (!CheckIfEligibleFreeTx(tx, view, stakes, pindex->nHeight, chainparams.GetConsensus())) {
+                    if (!CheckIfEligibleFreeTx(tx, view, stakes, pindex->nHeight, chainparams.GetConsensus(),
+                            fStakingWithdrawal || stakingTxParser.GetStakingTxType() == StakingTransactionType::BURN ||
+                            stakingTxParser.GetStakingTxType() == StakingTransactionType::DEPOSIT)) {
```
Comment (Jakub123Jakub123) [inline src/validation.cpp:2311]: TBH That quite ugly if

File: src/validation.cpp [hunk 2]
```diff
@@ -4501,7 +4512,10 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
         if (nCheckLevel >= 3 && (coins.DynamicMemoryUsage() + ::ChainstateActive().CoinsTip().DynamicMemoryUsage()) <= nCoinCacheUsage) {
             assert(coins.GetBestBlock() == pindex->GetBlockHash());
             if (pindex->nHeight >= chainparams.GetConsensus().nStakingStartHeight) {
-                assert(stakes.getBestBlock() == pindex->GetBlockHash());
+                if (stakes.getBestBlock() != pindex->GetBlockHash()) {
+                    return error("Found inconsistency in stakes DB. Restart with --reindex-chainstate flag to rebuild it.");
+                }
```
Comment (Jakub123Jakub123) [inline src/validation.cpp:4517]: Should be == or !=?
Assert was with ==, but now its different. Just checking

File: src/validation.cpp (hunk 1 above)
Comment (Jakub123Jakub123) [reply src/validation.cpp:2311]: Maybe more brackets with braking it over few lines would help to make it more readable

File: src/validation.cpp (hunk 2 above)
Comment (mtwaro) [reply src/validation.cpp:4517]: assert was supposed to throw on condition fail, now there is an error thrown on condition fulfilled.

**Review by Jakub123Jakub123 — APPROVED** [review]

**Pull request merged**